micronota/db/tigrfam.py

```diff
@@ -0,0 +1,236 @@
+
+import os
+import shutil
+import gzip
+import tarfile
+from tempfile import mkdtemp
+from os.path import join
+from ftplib import FTP
+from sqlite3 import connect
+
+from ..bfillings.hmmer import hmmpress_hmm
+
+
+def prepare_db(out_d, prefix='tigrfam_v15.0'):
+    '''Download and prepare TIGRFAM database.
+
+    Parameters
+    ----------
+    out_d : str
+        The directory of output files.
+    prefix : str
+        The file name (without extensions) of the output files.
+    '''
+    server = 'ftp.tigr.org'
+    path = 'pub/data/TIGRFAMs'
+
+    hmm = 'TIGRFAMs_15.0_HMM.LIB.gz'
+    hmm_out = join(out_d, '%s.hmm' % prefix)
+
+    metadata = 'TIGRFAMs_15.0_INFO.tar.gz'
+    metadata_out = join(out_d, '%s.db' % prefix)
+
+    try:
+        temp_dir = mkdtemp()
+        hmm_tmp = join(temp_dir, hmm)
+        metadata_tmp = join(temp_dir, metadata)
+        print(temp_dir)
+        with FTP(server) as ftp:
+            ftp.login()
+            ftp.cwd(path)
+
+            # fetch metadata file
+            ftp.retrbinary('RETR %s' % metadata,
+                           open(metadata_tmp, 'wb').write)
+            with tarfile.open(metadata_tmp) as tar:
+                tar.extractall(temp_dir)
+                prepare_metadata(temp_dir, metadata_out)
+
+            # fetch HMM model file
+            ftp.retrbinary('RETR %s' % hmm,
+                           open(hmm_tmp, 'wb').write)
+            with gzip.open(hmm_tmp, 'rb') as i_f, open(hmm_out, 'wb') as o_f:
+                shutil.copyfileobj(i_f, o_f)
+            hmmpress_hmm(hmm_out)
+    finally:
+        shutil.rmtree(temp_dir)
+
+
+def prepare_metadata(in_d, fp):
+    '''Compile the metadata into sqlite3 database.
+
+    Parameters
+    ----------
+    in_d : str
+        The input directory containing XXX.INFO files.
+    fp : str
+        The output file path of sqlite3 database.
+    '''
+    with connect(fp) as conn:
+        conn.execute("DROP TABLE IF EXISTS tigrfam")
+        conn.execute('''CREATE TABLE tigrfam (
+                            id       TEXT    NOT NULL,
+                            tag      TEXT    NOT NULL,
+                            value    BLOB    NOT NULL,
+                            transfer BOOLEAN NOT NULL,
+                        CHECK (transfer IN (0,1)));''')
+
+        for f in os.listdir(in_d):
+            if not f.endswith('.INFO'):
+                continue
+            tigrfam_id = f.split('.', 1)[0]
+            for x in _read_info(join(in_d, f)):
+                conn.execute('''INSERT INTO tigrfam (id, tag, value, transfer)
+                                    VALUES ("%s",?,?,?);''' % tigrfam_id,
+                             x)
+
+        conn.execute('CREATE INDEX id ON tigrfam (id);')
+        conn.commit()
+
+
+def _read_info(fn):
+    '''Parse the .INFO file.
+
+    Parameters
+    ----------
+    fn : str
+        file path
+
+    Yields
+    ------
+    tuple
+        tag, value, int of 0 or 1
```

Comment:
because sqlite3 doesn't have boolean type - it accepts 0 or 1 for that.

On Thu, Jan 21, 2016 at 7:47 PM, mortonjt <notifications@github.com> wrote:

> In micronota/db/tigrfam.py
> <https://github.com/biocore/micronota/pull/40#discussion_r50499603>:
>
> > +        conn.execute('CREATE INDEX id ON tigrfam (id);')
> > +        conn.commit()
> > +
> > +
> > +def _read_info(fn):
> > +    '''Parse the .INFO file.
> > +
> > +    Parameters
> > +    ----------
> > +    fn : str
> > +        file path
> > +
> > +    Yields
> > +    ------
> > +    tuple
> > +        tag, value, int of 0 or 1
>
> What is this int? You sure that you don't want bool?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/micronota/pull/40/files#r50499603>.
>